### PR TITLE
bpo-41249: Fix postponed annotations for TypedDict

### PR DESCRIPTION
**Pull request by Kronuz**
## Summary:

This fixes `TypedDict` to work with `get_type_hints()` and postponed evaluation of annotations across modules.

This is done by adding the module name to `ForwardRef` at the time the object is created and using that to resolve the globals during the evaluation.

## Test Plan:

Having these two file:

```py
# foo.py
from __future__ import annotations

from typing import Optional, TypedDict

class Foo(TypedDict):
    a: Optional[int]
```

```py
# bar.py
from __future__ import annotations

from typing import get_type_hints
from foo import Foo

class Bar(Foo, total=False):
    b: int

print(get_type_hints(Bar))
```

### Before:
```
$ python bar.py
Traceback (most recent call last):
  File "/Users/kronuz/Development/cpython/bar.py", line 9, in <module>
    print(get_type_hints(Bar))
  File "/Users/kronuz/Development/cpython/Lib/typing.py", line 1724, in get_type_hints
    value = _eval_type(value, base_globals, base_locals)
  File "/Users/kronuz/Development/cpython/Lib/typing.py", line 317, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/Users/kronuz/Development/cpython/Lib/typing.py", line 660, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'Optional' is not defined
```

### After:
```
$ python bar.py
{'a': typing.Optional[int], 'b': <class 'int'>}
```

https://bugs.python.org/issue41249

### CODE REVIEW

File: Lib/typing.py
Comment (Fidget-Spinner) [inline Lib/typing.py]: It's possible for `self.__forward_module__` to not be in `sys.modules` and raise `KeyError`. Maybe using this style would be better? https://github.com/will-ca/cpython/blob/090bf13b927e0eada702110806c0276f265e9584/Lib/typing.py#L1704

**Review by Fidget-Spinner — COMMENTED** [review]
Thanks for submitting this fix! Some general comments:
- Please add your proposed tests. Tests for typing are at https://github.com/python/cpython/blob/main/Lib/test/test_typing.py. Since your tests require multiple files, you can create them in that folder, or reuse this file that's used for testing type annotations in separate modules https://github.com/python/cpython/blob/main/Lib/test/mod_generics_cache.py)
- Please add a news entry via blurb.

I'll try to do a full review later.

**Comment by Kronuz** [comment]
@Fidget-Spinner, I added the tests, the news entry and fixed the `KeyError` as suggested. Thank you!

File: Lib/test/typed_dict.py
Comment (Fidget-Spinner) [inline Lib/test/typed_dict.py]: Please add a comment/docstring at the top describing the purpose of this file for future generations.

File: Misc/NEWS.d/next/Library/2021-07-04-11-33-34.bpo-41249.sHdwBE.rst
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2021-07-04-11-33-34.bpo-41249.sHdwBE.rst]: Usually technical implementation details are left out of the news article. The commit log is enough for that. WDYT?
```suggestion
```

**Review by Fidget-Spinner — COMMENTED** [review]
LG in general! I just have a few non-code comments. Thanks :).

File: Lib/test/typed_dict.py
Comment (Fidget-Spinner) [inline Lib/test/typed_dict.py]: bug number is enough, PR numbers are sometimes confusing because Python's gone through 3 or more version control systems at this point ;)

```suggestion
a `TypedDict` with postponed annotations, for bpo-41249.
```

File: Misc/NEWS.d/next/Library/2021-07-04-11-33-34.bpo-41249.sHdwBE.rst
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Library/2021-07-04-11-33-34.bpo-41249.sHdwBE.rst]: the double backticks format this as code
```suggestion
Fixes ``TypedDict`` to work with ``typing.get_type_hints()`` and postponed evaluation of
```

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Thanks for the quick response!

Please don't force push in the future. It overwrites valuable history, and besides, the core devs will squash and merge your PR as a single commit :).

I can't merge your PR, so I'll wait for Guido or some other core dev to review. If no one tends to this in a week or two, please gently ping the issue tracker.

**Comment by Kronuz** [comment]
@Fidget-Spinner, I made the suggested changes. Thank you for reviewing! 😃

File: Lib/test/typed_dict.py
Comment (gvanrossum) [inline Lib/test/typed_dict.py]: Can you rename the file to something like typed_dict_helper.py? I want it to be very clear that this is not seen as something exported by the test package. (Maybe even _typed_dict_helper.py?)

I also don't think that the bpo number needs to be mentioned in the docstring -- one can get that from the commit message. (It'd be lovely if the summary fit on one line, but I'm not holding you to that. :-)

**Review by gvanrossum — COMMENTED** [review]
LG except I'd like the helper module renamed. Thanks!

**Review by septatrix — COMMENTED** [review]
I still think it would be beneficial to figure out why TypedDicts have no MRO as this would trivially solve this problem. That being said I think this is a rather clean workaround for the time being

File: Lib/test/_typed_dict_helper.py
Comment (gvanrossum) [inline Lib/test/_typed_dict_helper.py]:
```suggestion
    class Bar(_typed_dict_helper.Foo, total=False):
```

**Review by gvanrossum — APPROVED** [review]
Just one nit, I’ll commit the fix and then land.

**Comment by Fidget-Spinner** [comment]
@Kronuz congrats on landing your first PR in CPython :)! Great work!

**Comment by gvanrossum** [comment]
Yes, good work!

**Comment by Fidget-Spinner** [comment]
I'll look into it.

**Comment by Fidget-Spinner** [comment]
Seems like bedevere was just feeling a little unwell ;).

**Pull request merged**
